### PR TITLE
debug_assert a few more raw pointer methods

### PR DESCRIPTION
**Pull request by RalfJung**
Makes progress for https://github.com/rust-lang/rust/issues/53871

### CODE REVIEW

**Comment by Mark-Simulacrum** [comment]
I believe we're already doing this elsewhere and these are consistent with that, so @bors r+ rollup
(probably in practice this is unlikely to cause breakage).

**Comment by JohnTitor** [comment]
This _likely_ caused the failure: https://github.com/rust-lang/rust/pull/69219#issuecomment-586801656
Could you confirm?

**Comment by RalfJung** [comment]
Hm, PR CI passed and I don't see what the error message would have to do with this PR. So, doesn't seem likely to me... but I have seen stranger things. Why do you think it was this one?

**Comment by JohnTitor** [comment]
I suspect this and #68767, but it's a tweak for macOS so I've commented here. Hm so likely later one caused the failure (i.e. this PR is unrelated)?

**Comment by JohnTitor** [comment]
Okay, next rollup contains this, if the failure isn't shown, I'll visit later one (sorry for the noise!).

**Comment by RalfJung** [comment]
No you were right, I can reproduce the failure locally.

@bors r-

I am at a loss though about what could be causing this...
Cc @arielb1 @michaelwoerister from https://github.com/rust-lang/rust/issues/40883 and Cc @hanna-kruppe 
Does anyone know how this PR could make [this test](https://github.com/rust-lang/rust/blob/75b98fbe77d472d85d1691bae5b25e7eefb3609c/src/test/ui/issues/issue-40883.rs) fail?

**Comment by Mark-Simulacrum** [comment]
Perhaps some function isn't getting inlined as much and we're spilling more to the stack? It would be helpful to get llvm ir before/after this PR, I think. @bors try so we'll have public artifacts

**Comment by RalfJung** [comment]
Removing the `debug_assert!` in `ptr::write` makes the test pass again.

Also, the computed stack usage is less than 100 bytes too big, so increasing the "slack" from 128 to 256 bytes would be sufficient to fix the test failure.

**Comment by RalfJung** [comment]
On my system, the patch I just added solves the problem. Not sure if that is acceptable, though.

**Comment by RalfJung** [comment]
@Mark-Simulacrum @cramertj does adjusting the size limit like I did not seem reasonable? The overall limit is still less than three times the size of `Big`, so we still know we don't get an extra copy of that struct somewhere on the stack.

**Comment by Mark-Simulacrum** [comment]
Oh, actually, that try build is useless as it didn't build with debug-assertions... never mind.

Yeah, I think increasing the slack space here is fine, I was hoping to dig in a bit more and figure out exactly what's changed but I don't think it's worth it, almost no one is going to be running with a debug-assert std anyway.

@bors r+

**Comment by eddyb** [comment]
@bors rollup=never (it accidentally stuck from before a failure was observed)

Potentially responsible for https://github.com/rust-lang/rust/pull/69463#issuecomment-591071983.

**Comment by RalfJung** [comment]
I ran codegen tests locally and they seem to work fine. Hm.

**Comment by eddyb** [comment]
@RalfJung The failure I linked is for the `arm-linux-androideabi` target specifically.
Not sure what about that could cause this (maybe 32-bit? then it might be possible to reproduce using `i686-unknown-linux-gnu`?).

It could even be some different configuration for that one builder (e.g. enabling debug assertions? even if I doubt it).

**Comment by RalfJung** [comment]
@eddyb I have enabled debug assertions locally. That was my first suspicion, but that's not it -- or at least, it is not alone sufficient. Maybe "debug assertions + 32bit", or so.

Now it happened on wasm32, it seams? I am somewhat out of my depth to debug this, I am afraid. We could just add `ignore-debug` like we [did for some other tests](https://github.com/rust-lang/rust/blob/46f5aa93d47e9077775ad9038970fd4c77abaad7/src/test/codegen/vec-clear.rs#L1), but that doesn't really explain what is happening.

File: src/test/ui/issues/issue-40883.rs [hunk 1]
```diff
@@ -78,8 +78,8 @@ fn verify_stack_usage(before_ptr: *mut Vec<Big>) {
     let stack_usage = isize::abs(
         (&mut stack_var as *mut _ as isize) -
             (before_ptr as isize)) as usize;
-    // give space for 2 copies of `Big` + 128 "misc" bytes.
-    if stack_usage > mem::size_of::<Big>() * 2 + 128 {
+    // give space for 2 copies of `Big` + 256 "misc" bytes.
+    if stack_usage > mem::size_of::<Big>() * 2 + 256 {
```
Comment (eddyb) [inline src/test/ui/issues/issue-40883.rs]: Hmm why would this change? I would hope your additions don't actually need `format_args!` but it's possible `debug_assert!` is inefficient?

File: src/libcore/intrinsics.rs [hunk 2]
```diff
@@ -1524,7 +1525,7 @@ pub unsafe fn copy_nonoverlapping<T>(src: *const T, dst: *mut T, count: usize) {
 
     debug_assert!(is_aligned_and_not_null(src), "attempt to copy from unaligned or null pointer");
     debug_assert!(is_aligned_and_not_null(dst), "attempt to copy to unaligned or null pointer");
-    debug_assert!(!overlaps(src, dst, count), "attempt to copy to overlapping memory");
+    debug_assert!(is_nonoverlapping(src, dst, count), "attempt to copy to overlapping memory");
```
Comment (eddyb) [inline src/libcore/intrinsics.rs]: Can you split this change into its own PR? That should have none of the problems this does.

Comment (RalfJung) [reply src/libcore/intrinsics.rs]: If we talk about splitting this, I this the best way forward is to remove the `write` change; I doubt `swap_nonoverlapping` is so critical.

Then I can submit a separate PR for the `write` assertion.

File: src/libcore/ptr/mod.rs [hunk 3]
```diff
@@ -795,6 +801,7 @@ pub unsafe fn read_unaligned<T>(src: *const T) -> T {
 #[inline]
 #[stable(feature = "rust1", since = "1.0.0")]
 pub unsafe fn write<T>(dst: *mut T, src: T) {
+    debug_assert!(is_aligned_and_not_null(dst), "attempt to write to unaligned or null pointer");
```
Comment (eddyb) [inline src/libcore/ptr/mod.rs]: I wonder if this is the problem, maybe try landing `swap_nonoverlapping` and this separately.

Maybe some uses of `ptr::write`  want to use `intrinsics::move_val_init` directly without creating a raw pointer? This is so important, IIRC we lower `intrinsics::move_val_init` in MIR building directly.

**Comment by RalfJung** [comment]
All right, I removed the assertion from `ptr::write`. That should fix the test failure.

Since this is a subset of the previous changes: @bors r=Mark-Simulacrum rollup-
